Sns doc commands (#1755)
dfinity/portal: docs/developer-docs/integrations/sns/managing/making-proposals.md

    };

    //proposals types for managing an SNS
    type Action = variant {
        ManageNervousSystemParameters : NervousSystemParameters;
        AddGenericNervousSystemFunction : NervousSystemFunction;
        RemoveGenericNervousSystemFunction : nat64;
        UpgradeSnsToNextVersion : record {};
        RegisterDappCanisters : RegisterDappCanisters;
        TransferSnsTreasuryFunds : TransferSnsTreasuryFunds;
        UpgradeSnsControlledCanister : UpgradeSnsControlledCanister;
        DeregisterDappCanisters : DeregisterDappCanisters;
        Unspecified : record {};
        ManageSnsMetadata : ManageSnsMetadata;
        ExecuteGenericNervousSystemFunction : ExecuteGenericNervousSystemFunction;
        Motion : Motion;
    };
```

## Using quill to submit proposals

### Requirements

* Installing [`quill`](https://github.com/dfinity/quill).
* principal that owns an SNS neuron that can make proposals for an SNS

### Submitting via `sns make-proposal` command

Any eligible neuron can submit a proposal. Therefore, the command to [submit a proposal](https://github.com/dfinity/quill/blob/master/docs/cli-reference/sns/quill-sns-make-proposal.md)

> `sns make-proposal` is a `ManageNeuron` message. With this command, neuron holders can submit proposals (such as a `Motion` Proposal) to be voted on by other neuron holders.

Reviewer: ```suggestion
`sns make-proposal` is a `ManageNeuron` message. With this command, neuron holders can submit proposals (such as a `Motion` proposal) to be voted on by other neuron holders.
```

The structure of the commands is as follows:

```bash
# create and sign the proposal, store it in a message.json file
quill sns --canister-ids-file <PATH_TO_CANISTER_IDS_JSON_FILE> --pem-file <PATH_TO_PEM_FILE> make-proposal <PROPOSAL_NEURON_ID> --proposal '(
    record {
        title = "lorem ipsum";
        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {
            <PROPOSAL_TYPE> = <PARAMETERS_OF_PROPOSAL_TYPE>
        };
    }
)' > message.json

# send the proposal (stored in message.json) to the network
quill send message.json
```

> * `<PATH_TO_CANISTER_IDS_JSON_FILE>` is the file path to a canister IDs JSON file. See example [sns_canister_ids.json](https://github.com/dfinity/quill/blob/master/e2e/assets/sns_canister_ids.json.)

Reviewer: ```suggestion
* `<PATH_TO_CANISTER_IDS_JSON_FILE>` is the file path to a canister IDs JSON file. See example [sns_canister_ids.json](https://github.com/dfinity/quill/blob/master/e2e/assets/sns_canister_ids.json).
```

> * `PROPOSAL_NEURON_ID` is the neuron ID of the neuron that is submitting the proposal

Reviewer: ```suggestion
* `PROPOSAL_NEURON_ID` is the neuron ID of the neuron that is submitting the proposal.
```

* `<PATH_TO_PEM_FILE>` is the path to the PEM file of the identity that owns the neuron that is submitting the proposal. To generate a PEM file, see [here](https://internetcomputer.org/docs/current/references/quill-cli-reference/quill-generate).

> 

Reviewer: ```suggestion
```

> * `title` is a short description of the proposal

Reviewer: ```suggestion
* `title` is a short description of the proposal.
```

> * `url` is a link to a document that describes the proposal in more detail

Reviewer: ```suggestion
* `url` is a link to a document that describes the proposal in more detail.
```

> * `summary` is a short summary of the proposal

Reviewer: ```suggestion
* `summary` is a short summary of the proposal.
```

> * `action` defines what the proposal does.. depending on the kind of proposal we require to provide different paramters that are defined in this part. As a proposal is just a call to a method, these parameters define with which arguments the target method will be called.

Reviewer: ```suggestion
* `action` defines what the proposal does. Depending on the kind of proposal we require to provide different parameters that are defined in this part. As a proposal is just a call to a method, these parameters define with which arguments the target method will be called.
```

export PROPOSAL_NEURON_ID="594fd5d8dce3e793c3e421e1b87d55247627f8a63473047671f7f5ccc48eda63"
# example path for the PEM file. This is a sample PEM file path.
export PEM_FILE="/home/user/.config/dfx/identity/$(dfx identity whoami)/identity.pem"

# Note: <PROPOSAL_TYPE> is replaced with "Motion" and <PARAMETERS_OF_PROPOSAL_TYPE> with the parameters for the Motion proposal
quill sns --canister-ids-file ./sns_canister_ids.json --pem-file $PEM_FILE make-proposal $PROPOSAL_NEURON_ID --proposal '(
    record {
        title = "SNS is great";
        url = "https://sns-examples.com/proposal/42";
        summary = "This is a motion proposal to see if people agree on the fact that the SNS is great.";
        action = opt variant {
            Motion = record {

                motion_text = "I hereby raise the motion that the use of the SNS shall commence";

            }        
        };
    }
)' > message.json

quill send message.json
```

:::warning
In this article, we will not repeat the `export PROPOSAL_NEURON_ID` and `export PEM_FILE` lines for each example proposal, but it is recommended you set these variables in your terminal before submitting proposals.
:::

## `Motion`

> A motion proposal is the only kind of proposal that does not have any immediate effect, i.e., it does not trigger the execution of a method as other proposals do. It can be used, for example for opinion polls before even starting certain features.  

Reviewer: ```suggestion
A motion proposal is the only kind of proposal that does not have any immediate effect, i.e., it does not trigger the execution of a method as other proposals do. For example, it can be used for opinion polls before even starting certain features.  
```

                id = 4_003 : nat64;
                name = "Import proposals group into community";
                description = opt "Import the specified proposals group into the specified community.";
                function_type = opt variant { 
                    GenericNervousSystemFunction = record { 

                        validator_canister_id = opt principal "iywa7-ayaaa-aaaaf-aemga-cai"; 

                        target_canister_id = opt principal "iywa7-ayaaa-aaaaf-aemga-cai"; 
                        
                        validator_method_name = opt "import_proposals_group_into_community_validate"; 
                        
                        target_method_name = opt "import_proposals_group_into_community";
                    } 
                };
            }
        };
    }
)' > message.json

quill send message.json
```

See example [proposal of an active SNS](https://dashboard.internetcomputer.org/sns/3e3x2-xyaaa-aaaaq-aaalq-cai/proposal/177).

## `RemoveGenericNervousSystemFunction`

Similarly to how generic proposal are added, they can also be removed again if the SNS DAO thinks that they should no longer be supported. 

> To do so, they can use the proposal  `AddGenericNervousSystemFunction` which specifies the ID of the generic nervous system function to be removed.

Reviewer: ```suggestion
To do so, they can use the proposal `AddGenericNervousSystemFunction` which specifies the ID of the generic nervous system function to be removed.
```

Each SNS community can then simply decide if and when they want to upgrade their SNS instance to the next SNS version that is available on SNS-W.

To do so, they can use an `UpgradeSnsToNextVersion`proposal. If this proposal is adopted, it will trigger a call to SNS root that will ask SNS-W which new wasm version is available and then upgrade to this new version.

### Relevant type signatures

```candid
    type UpgradeSnsToNextVersion : record {};
```

### Putting it together

Example in bash:

```bash
quill sns --canister-ids-file ./sns_canister_ids.json --pem-file $PEM_FILE make-proposal $PROPOSAL_NEURON_ID --proposal '(
    record {
        title = "Upgrade SNS to next available version";
        url = "https://sns-examples.com/proposal/42";
        summary = "A proposal to upgrade the SNS DAO to the next available version on SNS-W";
        action = opt variant {
            UpgradeSnsToNextVersion = record {};
        };
    }
)' > message.json

quill send message.json
```

## `RegisterDappCanisters`

An SNS controls a set of dapp cansiters. An SNS community can decide that new dapps should be added to the SNS' control.

> The proposal `RegisterDappCanisters`allows the SNS to accept the control of a set of new dapp canisters. The new canisters that should be registered are identified by their canister ID and it is allowed to register a list of canisters (not just a single one).

Reviewer: ```suggestion
The proposal `RegisterDappCanisters` allows the SNS to accept the control of a set of new dapp canisters. The new canisters that should be registered are identified by their canister ID and it is allowed to register a list of canisters (not just a single one).
```

### Relevant type signatures

```candid
    type DeregisterDappCanisters : DeregisterDappCanisters;

    type DeregisterDappCanisters = record {
        canister_ids : vec principal;
        new_controllers : vec principal;
    };
```

### Putting it together

Example in bash:

```bash
quill sns --canister-ids-file ./sns_canister_ids.json --pem-file $PEM_FILE make-proposal $PROPOSAL_NEURON_ID --proposal '(
    record {
        title = deregister dapp canisters";
        url = "https://sns-examples.com/proposal/42";
        summary = "This proposal gives up the control of a canister";
        action = opt variant {
            DeregisterDappCanisters = record {
                
                canister_ids = vec {principal "ltyfs-qiaaa-aaaak-aan3a-cai", principal "ltyfs-qiaaa-aaaak-aan3a-cai"};
                
                new_controllers = vec {principal "rymrc-piaaa-aaaao-aaljq-cai", principal "suaf3-hqaaa-aaaaf-bfyob-cai"};
            };
        };
    };
)' > message.json

quill send message.json
```

## `TransferSnsTreasuryFunds`

The SNS DAO has control over a treasury from which funds can be sent to other accounts by `TransferSnsTreasuryFunds` proposals.

> To do so, one has to define the following arguments... 

Reviewer: ```suggestion
To do so, one has to define the following arguments.
```

        summary = "Transfer 411 ICP to Foo Labs";
        action = opt variant {
            TransferSnsTreasuryFunds = record {

                from_treasury = 1 : int32;
                
                to_principal = opt principal "ozcnp-xcxhg-inakz-sg3bi-nczm3-jhg6y-idt46-cdygl-ebztx-iq4ft-vae";
                
                to_subaccount = null;
                
                memo = null;
                
                amount_e8s = 4_110_000_000_000 : nat64;
            };
        };
    };
)' > message.json

quill send message.json
```

See example [proposal of an active SNS](https://dashboard.internetcomputer.org/sns/3e3x2-xyaaa-aaaaq-aaalq-cai/proposal/202).

## `UpgradeSnsControlledCanister` 

The proposal `UpgradeSnsControlledCanister` is to upgrade a dapp canister that is controlled by the SNS DAO to a new wasm. 

### Relevant type signatures

```candid
    type UpgradeSnsControlledCanister : UpgradeSnsControlledCanister;

    type UpgradeSnsControlledCanister = record {
        new_canister_wasm : vec nat8;
        mode : opt int32;
        canister_id : opt principal;
        canister_upgrade_arg : opt vec nat8;
    };
```

:::warning

> Because this proposal requires passing Wasm which is unwieldly to copy/paste into the commandline as binary. It is recommended that developers use the specially-made [`make-upgrade-canister-proposal`](https://github.com/dfinity/quill/blob/master/docs/cli-reference/sns/quill-sns-make-upgrade-canister-proposal.md) command in `quill sns`.

Reviewer: ```suggestion
Because this proposal requires passing wasm, which is unwieldy to copy/paste into the command line as binary, it is recommended that developers use the specially-made [`make-upgrade-canister-proposal`](https://github.com/dfinity/quill/blob/master/docs/cli-reference/sns/quill-sns-make-upgrade-canister-proposal.md) command in `quill sns`.
```

```bash
quill sns make-upgrade-canister-proposal <PROPOSER_NEURON_ID> --target-canister-id <TARGET_CANISTER_ID> --wasm-path <WASM_PATH> [option]
```

```bash
export $WASM_PATH="/home/user/new_wasm.wasm"
quill sns make-upgrade-canister-proposal --target-canister-id "4ijyc-kiaaa-aaaaf-aaaja-cai" --wasm-path $WASM_PATH $PROPOSAL_NEURON_ID > message.json
quill send message.json
```

:::

## `ManageSnsMetadata`

> Each SNS has a metadata that defines the SNS project and which includes a URL, e.g., under which the main dapp canister can be found, a logo, a name, and a description summarizing the purpose of the projects. This metadata can be updated at any time, for example if there is a rebranding for the associated project. To do so, the SNS DAO can use the proposal  `ManageSnsMetadata` which defines all of the above attributes.

Reviewer: ```suggestion
Each SNS has metadata that defines the SNS project and includes a URL, e.g., under which the main dapp canister can be found, a logo, a name, and a description summarizing the purpose of the projects. This metadata can be updated at any time, for example if there is a rebranding for the associated project. To do so, the SNS DAO can use the proposal  `ManageSnsMetadata` which defines all of the below attributes.
```

    description : "Sample SNS used for educational purposes";
```

To update the `description` field, we can use the following command (where untouched fields get marked `null`).

Example in bash:

```bash
quill sns --canister-ids-file ./sns_canister_ids.json --pem-file $PEM_FILE make-proposal $PROPOSAL_NEURON_ID --proposal '(
    record {
        title = "lorem ipsum";
        url = "https://sns-examples.com/proposal/42";
        summary = "lorem ipsum";
        action = opt variant {
            ManageSnsMetadata = record {
                
                url = null;
                
                logo = null;
                
                name = null;
                
                description = "UPDATED Sample SNS used for educational purposes";
            };
        };
    }
)' > message.json

quill send message.json
```

Then the resulting metadata will end like this where only the `description` field changed:

```
    url = "https://sns-examples.com/proposal/42";
    logo : "https://sns-examples.com/logo/img.jpg";
    name : "SNS Example #42";
    description : "UPDATED Sample SNS used for educational purposes";
```

## `ExecuteGenericNervousSystemFunction` 

> After a generic proposal has been registered with a  `AddGenericNervousSystemFunction` proposal, such a proposal can be submitted with a `ExecuteGenericNervousSystemFunction` proposal.

Reviewer: ```suggestion
After a generic proposal has been registered with a `AddGenericNervousSystemFunction` proposal, such a proposal can be submitted with a `ExecuteGenericNervousSystemFunction` proposal.
```

There are the following types:

* `Motion`
* `ManageNervousSystemParameters`
* `AddGenericNervousSystemFunction`
* `RemoveGenericNervousSystemFunction`
* `UpgradeSnsToNextVersion`
* `RegisterDappCanisters`
* `TransferSnsTreasuryFunds`
* `UpgradeSnsControlledCanister`

> * `DeregisterDappCanisters`

Author: done

        subaccount : opt Subaccount 
    };

    //proposals types for managing an SNS
    type Action = variant {
        ManageNervousSystemParameters : NervousSystemParameters;
        AddGenericNervousSystemFunction : NervousSystemFunction;
        RemoveGenericNervousSystemFunction : nat64;
        UpgradeSnsToNextVersion : record {};
        RegisterDappCanisters : RegisterDappCanisters;
        TransferSnsTreasuryFunds : TransferSnsTreasuryFunds;
        UpgradeSnsControlledCanister : UpgradeSnsControlledCanister;
        DeregisterDappCanisters : DeregisterDappCanisters;
        Unspecified : record {};
        ManageSnsMetadata : ManageSnsMetadata;
        ExecuteGenericNervousSystemFunction : ExecuteGenericNervousSystemFunction;
        Motion : Motion;
    };
```

## Using quill to submit proposals

### Submitting via `sns make-proposal` command

`sns make-proposal` signs a `ManageNeuron` message to [submit a proposal](https://github.com/dfinity/quill/blob/master/docs/cli-reference/sns/quill-sns-make-proposal.md). With this command, neuron holders can submit proposals (such as a Motion Proposal) to be voted on by other neuron holders. The structure:

> 

Author: Agreed

> 

Author: done

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal <PROPOSAL> [option]
```

where `<PROPOSAL>` is a formatted a candid record:

```candid
(
    record {
        title = "lorem ipsum";
        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {
            <PROPOSAL_TYPE> = <PARAMETERS_OF_PROPOSAL_TYPE>
            }
        };
    }
)

> ```

Author: added it

The CLI command structure is

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {
        title = "lorem ipsum";
        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {
            <PROPOSAL_TYPE> = <PARAMETERS_OF_PROPOSAL_TYPE>
        };
    }
)'
```

For example, we use the candid record for <PROPOSAL_TYPE> `Motion`, the CLI-friendly command to submit a `Motion` proposal is:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {
        title = "SNS Launch";
        url = "https://dfinity.org\";
        summary = "A motion to start the SNS";
        action = opt variant {
            Motion = record {

>                 motion_text = "I hereby raise the motion that doesnt the use of the SNS shall commence";

Reviewer 2: ```suggestion
                motion_text = "I hereby raise the motion that the use of the SNS shall commence";
```

            }
        };
    }
)'
```

## References for proposals

### `Motion`

### Relevant Type signature

```candid
    Motion: record {
        motion_text : text;
    }
```

### Putting it together

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(

>     record {

Reviewer 2: would here use the exact same title / text etc as above or just refer to above and not repeat it

Author: agreed

        title = "lorem ipsum";
        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {
            Motion = record {
                motion_text = "I hereby raise the motion that the use of the SNS shall commence";
            }
        };
    }
)'
```

## `ManageNervousSystemParameters`

> 

Reviewer 2: ```suggestion
The nervous system parameters define the settings of a particular SNS. 
Since all SNSs are upgraded along the same upgrade path [refer to architecture part], this is what allows SNSs to nevertheless customise the rules of their governance, the tokenomics, etc to meet their needs. 

The proposal `ManageNervousSystemParameters` allows the SNS community to adjust these parameters at any time. 
```

> ### Relevant type signatures

Author: done

        neuron_minimum_stake_e8s : opt nat64;
        max_neuron_age_for_age_bonus : opt nat64;
        initial_voting_period_seconds : opt nat64;
        neuron_minimum_dissolve_delay_to_vote_seconds : opt nat64;
        reject_cost_e8s : opt nat64;
        max_proposals_to_keep_per_action : opt nat32;
        wait_for_quiet_deadline_increase_seconds : opt nat64;
        max_number_of_neurons : opt nat64;
        transaction_fee_e8s : opt nat64;
        max_number_of_proposals_with_ballots : opt nat64;
        max_age_bonus_percentage : opt nat64;
        neuron_grantable_permissions : opt record { permissions : vec int32 };
        voting_rewards_parameters : opt record {
            final_reward_rate_basis_points : opt nat64;
            initial_reward_rate_basis_points : opt nat64;
            reward_rate_transition_duration_seconds : opt nat64;
            round_duration_seconds : opt nat64;
        };
        maturity_modulation_disabled : opt bool;
        max_number_of_principals_per_neuron : opt nat64;
    }
```

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {

>         title = "Lorem ipsum";

Reviewer 2: ```suggestion
        title = "Increase the maximum dissolve delay";
```

How about we make one concrete example for what this proposal could be used?

>         url = "lorem ipsum";

Reviewer 2: example?

Author: yes thats the intent

>         summary = "lorem ipsum";

Reviewer 2: ```suggestion
        summary = "Proposal to adjust the SNS nervous system parameters so that newly the maximum dissolve delay that a neuron can have it set to 10 years. All other nervous system parameters should remain as they are now.";
```

        action = opt variant {
            ManageNervousSystemParameters: record {
                default_followees : opt record { 
                    followees : vec record { 
                                nat64; 

>                                 record { followees : vec record { id : vec {1:nat8, 2:nat8} } }  

Reviewer 2: Would this compile? I wonder: If I want to set one of these things to a new value, what would be the exact command that I need to use? Can I just plug in a number for nat64? Do I have to fill in all of these things, including those that should not change?
Again, I think a clear example would help

Author: I am not certain anything will compile yet, i have been doing these all by hand so far, but yes i intend to make sure all compiles before merging.

                initial_voting_period_seconds = opt 42:nat64;
                neuron_minimum_dissolve_delay_to_vote_seconds = opt 42:nat64;
                reject_cost_e8s = opt 42:nat64;
                max_proposals_to_keep_per_action = opt 42:nat32;
                wait_for_quiet_deadline_increase_seconds = opt 42:nat64;
                max_number_of_neurons = opt 42:nat64;
                transaction_fee_e8s = opt 42:nat64;
                max_number_of_proposals_with_ballots = opt 42:nat64;
                max_age_bonus_percentage = opt 42:nat64;
                neuron_grantable_permissions = opt record { permissions : vec {1:int32, 2:int32} };
                voting_rewards_parameters = opt record {
                    final_reward_rate_basis_points = opt 42:nat64;
                    initial_reward_rate_basis_points = opt 42:nat64;
                    reward_rate_transition_duration_seconds = opt 42:nat64;
                    round_duration_seconds = opt 42:nat64;
                };
                maturity_modulation_disabled = opt true:bool;
                max_number_of_principals_per_neuron = opt 42:nat64;
            };
    }
)'
```

> ## `AddGenericNervousSystemFunction`

Reviewer 2: I would separate this and the "SubmitGenericNervousSystemFunction" into a new section.

I would actually structure it like this (but we can also do this after example here are clear):
Background
- what are proposals (calls to methods..)
- how to submit a proposal (high level idea of quill, owning a neuron, what you need as prerequisites)

Native proposals
- explained on high level
- list all of them
- for each provide concrete example as now done here, for stuff relevant for generic proposals refer to next section

Generic proposals 
- what they are / how they work, including their structure (namely that they have an id but you cannot use certain ids as they are reserved for native proposals)
- how to add / remove generic proposal, including the example here (this will make the explanation here a lot easier)
- how to submit a generic proposal with example such as here

Author: Agreed, I just need some examples of generic proposals. will ask

Reviewer 2: Sent you a concrete example on slack (linked doc)

### Relevant type signatures

```candid
    type AddGenericNervousSystemFunction : NervousSystemFunction;

    type NervousSystemFunction = record {
        id : nat64;
        name : text;
        description : opt text;
        function_type : opt FunctionType;
    };

    type FunctionType = variant {
        NativeNervousSystemFunction : record {};
        GenericNervousSystemFunction : GenericNervousSystemFunction;
    };

    type GenericNervousSystemFunction = record {
        validator_canister_id : opt principal;
        target_canister_id : opt principal;
        validator_method_name : opt text;
        target_method_name : opt text;
    };
```

### Putting it together

```candid
    type AddGenericNervousSystemFunction: record {
        id : nat64;
        name : text;
        description : opt text;
        function_type : opt variant {

>             NativeNervousSystemFunction : record {};

Reviewer 2: since this is a variant, I would expect that we only have here GenericNervousSystemFunction or Native.. maybe good to double check with an engineer though
```suggestion

```

            GenericNervousSystemFunction : record {
                validator_canister_id : opt principal;
                target_canister_id : opt principal;
                validator_method_name : opt text;
                target_method_name : opt text;
            }
        };
    };
```

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {

>         title = "SNS Launch";

Reviewer 2: ```suggestion
        title = "Adding a new generic proposal";
```

        url = "https://dfinity.org";
        summary = "A motion to start the SNS";
        action = opt variant {
            AddGenericNervousSystemFunction = record {
                id = nat64;
                name = text;
                description = opt text;
                function_type : opt variant {
                    NativeNervousSystemFunction = record {};
                    GenericNervousSystemFunction = record {
                        validator_canister_id = opt principal "ltyfs-qiaab-aaaak-aan3a-cai";

>                         target_canister_id = opt principal;

Reviewer 2: I like that. I would fill in the same principal for the target_canister_id and then adjust the summary above accordingly to have one concrete example

Author: done

                        validator_method_name = opt text;
                        target_method_name = opt text;
                    }
                };
            }
        };
    }
)'
```

## `RemoveGenericNervousSystemFunction`

> 

Reviewer 2: ```suggestion
Similarly to how generic proposal are added, they can also be removed again if the SNS DAO thinks that they should no longer be supported. 
To do so, they can use the proposal  `AddGenericNervousSystemFunction` which specifies the ID of the generic nervous system function to be removed.
```

### Relevant type signatures

```candid
    type RemoveGenericNervousSystemFunction : nat64;
```

### Putting it together

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {
        title = "SNS Launch";
        url = "https://dfinity.org";
        summary = "A motion to start the SNS";
        action = opt variant {
            RemoveGenericNervousSystemFunction = 123456789:nat64;
        };
    }
)'
```

## `UpgradeSnsToNextVersion`

> 

Reviewer 2: ```suggestion
Recall that all approved SNS canister versions are stored on the NNS canister [SNS-W](introduction/sns-architecture). 
New SNS canister wasm codes are approved by NNS proposals and then added to SNS-W. 
Each SNS community can then simply decide if and when they want to upgrade their SNS instance to the next SNS version that is available on SNS-W. 
To do so, they can use an `UpgradeSnsToNextVersion`proposal. 
If this proposal is adopted, it will trigger a call to SNS root that will ask SNS-W which new wasm version is available and then upgrade to this new version. 
```

I think eventually it would be good to define this in more detail, but better to have already something now!

### Relevant type signatures

```candid
    type UpgradeSnsToNextVersion : record {};
```

### Putting it together

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {

>         title = "SNS Launch";

Reviewer 2: ```suggestion
        title = "Upgrade SNS to next available version";
```

        url = "https://dfinity.org";

>         summary = "A motion to start the SNS";

Reviewer 2: ```suggestion
        summary = "A proposal to upgrade the SNS DAO to the next available version on SNS-W";
```

        action = opt variant {
            UpgradeSnsToNextVersion = record {};
        };
    }
)'
```

## `RegisterDappCanisters`

> 

Reviewer 2: ```suggestion
An SNS controls a set of dapp cansiters. 
An SNS community can decide that new dapps should be added to the SNS' control. 
The proposal `RegisterDappCanisters`allows the SNS to accept the control of a set of new dapp canisters. The new canisters that should be registered are identified by their canister ID and it is allowed to register a list of canisters (not just a single one).
```

### Relevant type signatures

```candid
    type RegisterDappCanisters : RegisterDappCanisters;

    type RegisterDappCanisters = record { canister_ids : vec principal };
```

### Putting it together

```candid
    type RegisterDappCanisters: record {
        canister_ids : vec principal
    };
```

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {

>         title = "SNS Launch";

Reviewer 2: ```suggestion
        title = "Register new dapp canisters";
```

        url = "https://dfinity.org";
        summary = "A motion to start the SNS";
        action = opt variant {
            RegisterDappCanisters = record {
                canister_ids = vec {principal "ltyfs-qiaaa-aaaak-aan3a-cai", principal "ltyfs-qiaaa-aaaak-aan3a-cai"};
            };
        };
    }
)'
```

## `TransferSnsTreasuryFunds`

> 

Reviewer 2: ```suggestion
The SNS DAO has control over a treasury from which funds can be sent to other accounts by `TransferSnsTreasuryFunds` proposals. 
To do so, one has to define the following arguments... 
```

(I am not sure if "from_treasury" defines a ledger canister. I have to clarify this, but think it would be good to list all argument and what they mean here)

### Relevant type signatures

```candid
    type TransferSnsTreasuryFunds = record {
        from_treasury : int32;
        to_principal : opt principal;
        to_subaccount : opt Subaccount;
        memo : opt nat64;
        amount_e8s : nat64;
    };

    type Subaccount = record { subaccount : vec nat8 };
```

### Putting it together

```candid
    type TransferSnsTreasuryFunds = record {
        from_treasury : int32;
        to_principal : opt principal;
        to_subaccount : opt record { subaccount : vec nat8 };
        memo : opt nat64;
        amount_e8s : nat64;
    };
```

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {

>         title = "Lorem ipsum";

Reviewer 2: would again fill in a representative title / summary

Author: agreed. intent is to replace all the lorem ipsums

        url = "Lorem ipsum";
        summary = "Lorem ipsum";
        action = opt variant {
            TransferSnsTreasuryFunds = record {

>                 from_treasury = int32;

Reviewer 2: would fill in a concrete example, e.g., to move sth from the ICP treasury to a concrete subaccount and explain in the example taht this is what this does and why

                to_principal = opt principal;
                to_subaccount = opt record { subaccount = vec nat8 };
                memo = opt nat64;
                amount_e8s = nat64;
               
            };
        };
    };
)'
```

## `UpgradeSnsControlledCanister` 

> 

Reviewer 2: ```suggestion
The proposal `UpgradeSnsControlledCanister` is to upgrade a dapp canister that is controlled by the SNS DAO to a new wasm. 
```
(TODO: add explanation of all the arguments)

### Relevant type signatures

```candid
    type UpgradeSnsControlledCanister : UpgradeSnsControlledCanister;

    type UpgradeSnsControlledCanister = record {
        new_canister_wasm : vec nat8;
        mode : opt int32;
        canister_id : opt principal;
        canister_upgrade_arg : opt vec nat8;
    };
```

### Putting it together

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {
        title = "lorem ipsum";
        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {
            UpgradeSnsControlledCanister = record {
                new_canister_wasm = vec nat8;
                mode = opt int32;
                canister_id = opt principal;
                canister_upgrade_arg = opt vec nat8;
            };
        };
    };
)'
```

## `DeregisterDappCanisters` 

> 

Reviewer 2: ```suggestion
The proposal `DeregisterDappCanisters` is the counterpart of `RegisterDappCanisters`.
If an SNS community decides that they would like to give up the control of a given dapp canister, they can use this proposal to do so. 
To this end, the proposal defines the canister ID of the dapp canister to be deregistered and a principal to whom the canister will be handed over to (i.e., this principal will be set as the new controller of the specified dapp canister).
```

### Relevant type signatures

```candid
    type DeregisterDappCanisters : DeregisterDappCanisters;

    type DeregisterDappCanisters = record {
        canister_ids : vec principal;
        new_controllers : vec principal;
    };
```

### Putting it together

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {

>         title = "lorem ipsum";

Reviewer 2: would add concrete example

        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {
            DeregisterDappCanisters = record {

>                 canister_ids = vec principal;

Reviewer 2: would add concrete example

                new_controllers = vec principal;
            };
        };
    };
)'
```

> ### Putting it together
> 
> Example in bash:
> 
> ```bash
> quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
>     record {
>         title = "lorem ipsum";
>         url = "lorem ipsum";
>         summary = "lorem ipsum";
>         action = opt variant {
>             <PROPOSAL_TYPE> = <PARAMETERS_OF_PROPOSAL_TYPE>
>         };
>     }
> )'
> ```

Reviewer 2: ```suggestion

```
(think this is an accidential copy of the above)

## `ManageSnsMetadata` 

> 

Reviewer 2: ```suggestion
Each SNS has a metadata that defines the SNS project and which includes a URL, e.g., under which the main dapp canister can be found, a logo, a name, and a description summarizing the purpose of the projects.
This metadata can be updated at any time, for example if there is a rebranding for the associated project.
To do so, the SNS DAO can use the proposal  `ManageSnsMetadata` which defines all of the above attributes.
```

I would add: for those things that I don't want to change, do I need to leave it empty or do I need to specify all things including those that I don't want to change?

### Relevant type signatures

```candid
    type  ManageSnsMetadata : ManageSnsMetadata;

    type ManageSnsMetadata = record {
        url : opt text;
        logo : opt text;
        name : opt text;
        description : opt text;
    };
```

### Putting it together

> Example in bash:

Reviewer 2: would make a concrete example here and say "before the proposal it was like this..." 
the proposal now only cahnges 2 of the 4 attributes to this...
To do so, you could send this proposal..

(then it is clear how to define also the things that one does not want to change)

Author: Sorry, I am not sure I follow

Reviewer 2: sorry if this was unclear.
so this is a proposal to change the metadata of an SNS.
I thought it might be useful to talk the reader through a concrete example and say "let's say today the metadata is like this (e.g., state a name and description and url and logo). Then, we can make an example and say "let's say you now want to change the description, but leave the url, logo, and name as they are" and then show a concrete proposal that would change only this one field. 
My intent: sometimes it is also hard to know how to specify which fields you _don't_ want to change: do I leave them blank? do I need to just repeat the same value that is already there to say "I want to keep them"? => therefore I think an example that includes both things that an SNS wants to change and things that an SNS does not want to change might be super helpful

Author: done

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {
        title = "lorem ipsum";
        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {
            <PROPOSAL_TYPE> = <PARAMETERS_OF_PROPOSAL_TYPE>
        };
    }
)'
```

## `ExecuteGenericNervousSystemFunction` 

> 

Reviewer 2: ```suggestion
After a generic proposal has been registered with a  `AddGenericNervousSystemFunction` proposal, such a proposal can be submitted with a `ExecuteGenericNervousSystemFunction` proposal.
The proposal identifies the previously added generic proposal by an ID (so called `function_id`) and, in addition, defines a payload.
Upon submission of such a proposal, the defined validation method is called, which checks that the given payload is valid for this kind of proposal (the method that will be called for this is the method `validator_method_name` on the canister `validator_canister_id` as it was defined when the generic proposal was added. If this validation is successful, the proposal will be created.

Later, if the proposal is adopted, the SNS governance canister will call the method `target_method_name ` on the canister `target_canister_id` (as also define in the generic proposal) with the payload defined here.
```

Due to this I think it would be helpful to have the proposals for adding, removing, and executing generic proposals together. It will then make sense what each of the fields mean. Maybe there could even be one part "relevant type signatures" that summarizes it all.

### Relevant type signatures

```candid
    type ExecuteGenericNervousSystemFunction : ExecuteGenericNervousSystemFunction;

    type ExecuteGenericNervousSystemFunction = record {
        function_id : nat64;
        payload : vec nat8;
    };
```

### Putting it together

Example in bash:

```bash
quill sns make-proposal <PROPOSER_NEURON_ID> --proposal '(
    record {
        title = "lorem ipsum";
        url = "lorem ipsum";
        summary = "lorem ipsum";
        action = opt variant {

>             ExecuteGenericNervousSystemFunction = record {

Reviewer 2: would make conrete example
